src/libraries/System.Net.Security/src/System/Net/Security/SecureChannel.cs

```diff
@@ -1263,11 +1277,30 @@ private void LogCertificateValidation(RemoteCertValidationCallback remoteCertVal
     }
 
     // ProtocolToken - used to process and handle the return codes from the SSPI wrapper
-    internal class ProtocolToken
+    internal class ProtocolToken : IDisposable
     {
         internal SecurityStatusPal Status;
-        internal byte[] Payload;
         internal int Size;
+        private byte[] _payload;
+
+        internal byte[] Payload
+        {
+            get
+            {
+               return _payload;
+            }
+
+            set
+            {
+                if (SslStreamPal.RentsBufferForHanshake && _payload != null)
+                {
+                    ArrayPool<byte>.Shared.Return(_payload);
```

Comment:
Ooh, yeah, this definitely makes me nervous.  Erroneous setting of Payload could now result in the same buffer being returned multiple times to the pool, e.g. if we accidentally access it concurrently.  And it'll be worse if we do make ProtocolToken a struct (maybe that's why you chose not to).

Does _payload need to actually be a separate buffer?  We already have _internalBuffer on the SslStream... can we just use that?

Reply:
I was wondering about it but the ProtocolToken use is currently pretty simple and controlled by us. I originally did not have it IDisposable and I was worrying about loosing rented buffer in error cases and causing memory leaks. I call Reset() when we are done with token to release buffer as soon as possible. We can possibly assert or throw if buffer is not null when assigning new value. That would make sure we don't mingle usages. 

ProtocolToken is IMHO convenience wrapper and we could possibly abandon it and track status and buffers separately. That would make it easier to track life cycle. 


I was also thinking about using _internalBuffer. Handshake happens before we would normally use it and we rent buffer big enough for whole TLS record. The trouble would be handling renegotiation. It still may be possible but the _internalBuffer could have chunk of data while we would start new handshake. We could use same strategy on separate variable and keep the rented buffer in class variable instead of holding to it via ProtocolToken.

Let me know  what seems most appropriate.